### ISSUE TITLE
KHR_technique_webgl and TEXCOORD_0/1 type

### ISSUE THREAD

**Issue by stevesan** [KhronosGroup/glTF#1188]
One issue: glTF2 appears to limit TEXCOORD_0/1 to be VEC2's. However, if this extension is meant to support custom shaders, then it's very common for custom shaders to interpret UVs in different ways, and thus use VEC3/4 for UVs as well. As a part of this extension, should TEXCOORD_0/1 be allowed to have all 3 types?

**Comment by pjcozzi** [comment]
Good point.  The extension - or another extension or app - could introduce a new semantic, e.g., `TEXCOORD3_0`, but I don't think an extension should be allowed to expand the scope of an existing semantic.

**Comment by donmccurdy** [comment]
> ...it's very common for custom shaders to interpret UVs in different ways, and thus use VEC3/4 for UVs as well.

Are these attributes actually UVs in a real sense? If the attributes are being used for arbitrary data, I'd expect this line in the specification to apply:

> Application-specific semantics must start with an underscore, e.g., `_TEMPERATURE`.

**Comment by snagy** [comment]
>Application-specific semantics must start with an underscore, e.g., _TEMPERATURE.

Will this work with most authoring tools/pipelines?  I would guess most of our glTF exporters will support the typical semantics and not arbitrary stuff, but these exporters will have to be aware of the custom semantics in a shader.  Is that normal?

**Comment by donmccurdy** [comment]
My concern with using TEXCOORD_0/1 was that some engines may have hardwired assumptions about UVs being vec2, like [this example in THREE.BufferGeometry](https://github.com/mrdoob/three.js/blob/dev/src/core/BufferGeometry.js#L446) or these results in the [BabylonJS codebase](https://github.com/BabylonJS/Babylon.js/search?utf8=%E2%9C%93&q=uv&type=).

Not opposed to loosening the UV requirements if we find authoring tools that can't get around this, but would be nice to have a concrete example before doing so.

**Comment by stevesan** [comment]
The specific use case I'm concerned with is Tilt Brush sketches. TB uses texture coordinates for many purposes and tricks. So absolutely, the exporter is tightly coupled with the shader, and both are aware of the custom semantics.

I'll take some time to see if custom semantics. The idea of *adding* a semantic, like TEXCOORD3_0, is interesting...but yeah maybe it's pointless because what would an implementation even do with that information?

**Comment by stevesan** [comment]
OK I tried just turning all TEXCOORD_0/1 semantics into _CUSTOM_TEXCOORD_0/1. Using three.js's legacy glTF1.0 loader (which I'm assuming will be used as the basis for three.js's implementation of this extension), this doesn't work, because the code only does anything for the known semantics: https://github.com/mrdoob/three.js/blob/r86/examples/js/loaders/GLTFLoader.js#L1624 Otherwise, the attribute is just dropped. Should this be considered a bug (just hypothetically speaking, since we'll be using the glTF2.0 loader anyway)? Should KHR_technique_webgl be more specific about the handling of custom attributes, or at least include a conformance test case?

**Comment by lexaknyazev** [comment]
Extensions should be able to introduce their own standardized semantics and/or extend existing ones. We don't have an example of the latter case, though.

Imo, as long as extension is mentioned in `extensionsRequired`, it could "redefine" existing semantics as it needs. Incompatible loaders will reject such assets anyway.

On the other hand, if there's any value in having a fallback behavior , extensions should use custom semantic.

**Comment by donmccurdy** [comment]
> Should this be considered a bug (just hypothetically speaking, since we'll be using the glTF2.0 loader anyway)?

Can be considered a bug I think, no reason we shouldn’t be able to fully support custom semantics in the three.js glTF 2.0 loader. Would be nice to put some examples in the sample repo when the extension is ready, at a minimum.

**Comment by stevesan** [comment]
OK, I modified the glTF 1.0 loader to support custom semantics, and that seems to work. Standard semantics are mapped to three.js's preferred names (like TEXCOORD_0 to uv, TEXCOORD_1 to uv2, etc.), and shader code is edited to use those names. So for custom semantics, I'm just re-using the semantic name as the attribute name. The assumption is that they are unique within a technique's attributes.

So in conclusion, a custom semantic works fine, but we should:
1. Have a conformance test/example to make sure they're implemented properly.
2. Make it clear that they must be unique within a technique.

Anything else I can do to move this along? Thanks much.

**Comment by stevesan** [comment]
I'm going to close this issue for now. Strictly speaking, the spec does say something about application-specific semantics, so that should suffice. It could be a bit more specific, and a conformance test would be nice, but that warrants a separate issue.
